pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,368 @@
+type ImageRepo struct {
+	FullName string       `json:"fullname"`
+	Tags     []string     `json:"tags,omitempty"`
+	Edges    []*Edge      `json:"edges,omitempty"`
+	Children []*ImageRepo `json:"children,omitempty"`
+}
+
+// String helps in dumping a tree in AST format
+func (root *ImageRepo) String() string {
+	tree := ""
+	tree += root.FullName
+
+	for _, n := range root.Children {
+		tree += "(" + n.String() + ")"
+	}
+
+	return tree
+}
+
+// Edge represents a build configuration relationship
+// between two nodes
+//
+// Note that this type has no relation with the dot.Edge
+// type
+type Edge struct {
+	FullName string `json:"fullname"`
+	To       string `json:"to"`
+}
+
+// NewEdge adds a new edge on a parent node
+//
+// Note that this function has no relation
+// with the dot.Edge type
+func NewEdge(fullname, to string) *Edge {
+	return &Edge{
+		FullName: fullname,
+		To:       to,
+	}
+}
+
+// NewCmdBuildChain implements all the necessary functionality for the build-chain module
+func NewCmdBuildChain(f *clientcmd.Factory, parentName, name string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s [namespace]/[image-repository]:[tag] [OPTIONS]", name),
+		Short: "Output build dependencies of a specific image repository",
+		Long:  longDescription,
+		Run: func(cmd *cobra.Command, args []string) {
+			all := cmdutil.GetFlagBool(cmd, "all")
+			if len(args) > 1 || len(args) == 1 && all {
+				usageError(cmd, "Must pass nothing, an image repository name:tag combination, or specify the --all flag")
+			}
+
+			osc, kc, err := f.Clients(cmd)
+			checkErr(err)
+
+			// Retrieve namespace(s)
+			defaultNamespace, err := f.DefaultNamespace(cmd)
+			checkErr(err)
+			namespaces := make([]string, 0)
+			switch all {
+			case true:
+				nsList, err := kc.Namespaces().List(labels.Everything())
+				checkErr(err)
+				for _, ns := range nsList.Items {
+					namespaces = append(namespaces, ns.Name)
+				}
+			case false:
+				if len(args) != 0 {
+					namespace, _, _, err := parse(args[0], defaultNamespace)
+					checkErr(err)
+					defaultNamespace = namespace
+				}
+				namespaces = append(namespaces, defaultNamespace)
+			}
+
+			// Get all build configurations
+			buildConfigList := make([]buildapi.BuildConfig, 0)
+			for _, namespace := range namespaces {
+				cfgList, err := osc.BuildConfigs(namespace).List(labels.Everything(), labels.Everything())
+				checkErr(err)
+				buildConfigList = append(buildConfigList, cfgList.Items...)
+			}
+
+			// Get all image repositories from build configurations
+			repos := getRepos(buildConfigList)
+
+			// Parse user input and validate possible specified image repository
+			if !all && len(args) != 0 {
+				namespace, name, tag, err := parse(args[0], defaultNamespace)
+				checkErr(err)
+
+				// Validate the specified image repository
+				repo := join(namespace, name)
+				if _, ok := repos[repo]; !ok {
+					checkErr(fmt.Errorf("invalid image repository: %s", repo))
```

Comment:
it isn't really invalid, it's just an image repo that doesn't trigger any builds, right?